File: problems-extend/0901-1000/0984_string-without-aaa-or-bbb/solution.go
import "strings"

// The judge pins one exact answer: call the letter with the larger count
// big ('a' on a tie) and the other small. While big exceeds small and small
// has not run out, append two big letters then one small letter; then, while
// letters remain, append one big letter if any are left, then one small
// letter if any are left.
func strWithout3a3b(a int, b int) string {
	big, bigLetter := a, byte('a')
	small, smallLetter := b, byte('b')
	if b > a {
		big, bigLetter = b, byte('b')
		small, smallLetter = a, byte('a')
	}
	var answer strings.Builder
	for big > small && small > 0 {
		answer.WriteByte(bigLetter)
		answer.WriteByte(bigLetter)
		answer.WriteByte(smallLetter)
		big -= 2
		small--
	}
	for big > 0 || small > 0 {
		if big > 0 {
			answer.WriteByte(bigLetter)
			big--
		}
		if small > 0 {
			answer.WriteByte(smallLetter)
			small--
		}
	}
	return answer.String()
}

File: problems-extend/0901-1000/0984_string-without-aaa-or-bbb/solution.ts
// The judge pins one exact answer: call the letter with the larger count
// big ('a' on a tie) and the other small. While big exceeds small and small
// has not run out, append two big letters then one small letter; then, while
// letters remain, append one big letter if any are left, then one small
// letter if any are left.
function strWithout3a3b(a: number, b: number): string {
    let big = a;
    let bigLetter = "a";
    let small = b;
    let smallLetter = "b";
    if (b > a) {
        big = b;
        bigLetter = "b";
        small = a;
        smallLetter = "a";
    }
    const parts: string[] = [];
    while (big > small && small > 0) {
        parts.push(bigLetter + bigLetter + smallLetter);
        big -= 2;
        small -= 1;
    }
    while (big > 0 || small > 0) {
        if (big > 0) {
            parts.push(bigLetter);
            big -= 1;
        }
        if (small > 0) {
            parts.push(smallLetter);
            small -= 1;
        }
    }
    return parts.join("");
}
